lib/base/tlsutility.cpp

```diff
@@ -760,24 +761,27 @@ std::shared_ptr<X509> CreateCertIcingaCA(const std::shared_ptr<X509>& cert)
 	return CreateCertIcingaCA(pkey.get(), X509_get_subject_name(cert.get()));
 }
 
-bool IsCertUptodate(const std::shared_ptr<X509>& cert)
+bool IsCertUptodate(X509* cert)
 {
 	time_t now;
 	time(&now);
 
+	bool isRootCa = !X509_NAME_cmp(X509_get_subject_name(cert), X509_get_issuer_name(cert));
+
 	/* auto-renew all certificates which were created before 2017 to force an update of the CA,
 	 * because Icinga versions older than 2.4 sometimes create certificates with an invalid
 	 * serial number. */
 	time_t forceRenewalEnd = 1483228800; /* January 1st, 2017 */
-	time_t renewalStart = now + RENEW_THRESHOLD;
+	time_t renewalStart = now + (isRootCa ? std::max(RENEW_THRESHOLD, LEAF_VALID_FOR) : RENEW_THRESHOLD);
 
-	return X509_cmp_time(X509_get_notBefore(cert.get()), &forceRenewalEnd) != -1 && X509_cmp_time(X509_get_notAfter(cert.get()), &renewalStart) != -1;
+	return (isRootCa || X509_cmp_time(X509_get_notBefore(cert), &forceRenewalEnd) != -1)
+		&& X509_cmp_time(X509_get_notAfter(cert), &renewalStart) != -1;
```

Comment:
Completely re-done this and re-cherry-picked the whole branch, but the lib/ diff is exactly as needed 👍:

```patch
diff --git a/lib/base/tlsutility.cpp b/lib/base/tlsutility.cpp
index cea0738a4..03dd4c95e 100644
--- a/lib/base/tlsutility.cpp
+++ b/lib/base/tlsutility.cpp
@@ -8,7 +8,6 @@
 #include "base/utility.hpp"
 #include "base/application.hpp"
 #include "base/exception.hpp"
-#include <algorithm>
 #include <boost/asio/ssl/context.hpp>
 #include <openssl/opensslv.h>
 #include <openssl/crypto.h>
@@ -761,21 +760,27 @@ std::shared_ptr<X509> CreateCertIcingaCA(const std::shared_ptr<X509>& cert)
        return CreateCertIcingaCA(pkey.get(), X509_get_subject_name(cert.get()));
 }

-bool IsCertUptodate(X509* cert)
+static inline
+bool IsCertExpiring(X509* cert, int threshold)
 {
-       time_t now;
-       time(&now);
+       time_t renewalStart = time(nullptr) + threshold;

-       bool isRootCa = !X509_NAME_cmp(X509_get_subject_name(cert), X509_get_issuer_name(cert));
+       return X509_cmp_time(X509_get_notAfter(cert), &renewalStart) < 0;
+}

+bool IsCertUptodate(const std::shared_ptr<X509>& cert)
+{
        /* auto-renew all certificates which were created before 2017 to force an update of the CA,
         * because Icinga versions older than 2.4 sometimes create certificates with an invalid
         * serial number. */
        time_t forceRenewalEnd = 1483228800; /* January 1st, 2017 */
-       time_t renewalStart = now + (isRootCa ? std::max(RENEW_THRESHOLD, LEAF_VALID_FOR) : RENEW_THRESHOLD);

-       return (isRootCa || X509_cmp_time(X509_get_notBefore(cert), &forceRenewalEnd) != -1)
-               && X509_cmp_time(X509_get_notAfter(cert), &renewalStart) != -1;
+       return X509_cmp_time(X509_get_notBefore(cert.get()), &forceRenewalEnd) != -1 && !IsCertExpiring(cert.get(), RENEW_THRESHOLD);
+}
+
+bool IsCaUptodate(X509* cert)
+{
+       return !IsCertExpiring(cert, LEAF_VALID_FOR);
 }

 String CertificateToString(X509* cert)
diff --git a/lib/base/tlsutility.hpp b/lib/base/tlsutility.hpp
index 5bf15b7a1..b06412020 100644
--- a/lib/base/tlsutility.hpp
+++ b/lib/base/tlsutility.hpp
@@ -68,12 +68,8 @@ inline String CertificateToString(const std::shared_ptr<X509>& cert)
 std::shared_ptr<X509> StringToCertificate(const String& cert);
 std::shared_ptr<X509> CreateCertIcingaCA(EVP_PKEY *pubkey, X509_NAME *subject, bool ca = false);
 std::shared_ptr<X509> CreateCertIcingaCA(const std::shared_ptr<X509>& cert);
-bool IsCertUptodate(X509* cert);
-
-inline bool IsCertUptodate(const std::shared_ptr<X509>& cert)
-{
-       return IsCertUptodate(cert.get());
-}
+bool IsCertUptodate(const std::shared_ptr<X509>& cert);
+bool IsCaUptodate(X509* cert);

 String PBKDF2_SHA1(const String& password, const String& salt, int iterations);
 String PBKDF2_SHA256(const String& password, const String& salt, int iterations);
diff --git a/lib/remote/apilistener.cpp b/lib/remote/apilistener.cpp
index 82168545b..85443e218 100644
--- a/lib/remote/apilistener.cpp
+++ b/lib/remote/apilistener.cpp
@@ -339,7 +339,7 @@ void ApiListener::RenewCA()
        auto certPath (GetCaDir() + "/ca.crt");
        auto cert (GetX509Certificate(certPath));

-       if (IsCertUptodate(cert)) {
+       if (IsCaUptodate(cert.get())) {
                return;
        }

diff --git a/lib/remote/jsonrpcconnection-pki.cpp b/lib/remote/jsonrpcconnection-pki.cpp
index 12aea99ae..340e12b30 100644
--- a/lib/remote/jsonrpcconnection-pki.cpp
+++ b/lib/remote/jsonrpcconnection-pki.cpp
@@ -112,7 +112,7 @@ Value RequestCertificateHandler(const MessageOrigin::Ptr& or
igin, const Dictiona
                                }
                        }

-                       if (requestorCA && !IsCertUptodate(requestorCA)) {
+                       if (requestorCA && !IsCaUptodate(requestorCA)) {
                                int days;

                                if (ASN1_TIME_diff(&days, nullptr, X509_get_notAfter(requestorCA), X509_get_notAfter(cacert.get())) && days > 0) {```